### ISSUE TITLE
Is it possible to distinguish between absent (no value) and null when serializing?

### ISSUE THREAD

**Issue by GP4cK** [google/built_value.dart#1123]
Let's say I have a SimpleValue class with a nullableString:

```dart
abstract class SimpleValue implements Built<SimpleValue, SimpleValueBuilder> {
  @BuiltValueSerializer(serializeNulls: true)
  static Serializer<SimpleValue> get serializer => _$simpleValueSerializer;

  String? get nullableString;

  factory SimpleValue([void Function(SimpleValueBuilder) updates]) = _$SimpleValue;
  SimpleValue._();
}
```

I would like to be able to have this behaviour when serializing to JSON:

```dart
final noValue = SimpleValue();
print(serializers.toJson(SimpleValue.serializer, value)); // should print {}

final nullValue = SimpleValue((value) => value.nullableString = null);
print(serializers.toJson(SimpleValue.serializer, nullValue)); // should print { "nullableString": null }
```

Is it possible?

**Comment by davidmorgan** [comment]
Please try tagging your `Serializer` getter with

https://pub.dev/documentation/built_value/latest/built_value/BuiltValueSerializer-class.html

and setting `serializeNulls: true`, that should do it :)

**Comment by GP4cK** [comment]
Yes I tried that (it's already in the code I posted above). However it will always print `{ "nullableString": null }`. Instead, I would like that if I don't explicitly set `nullableString` to null when creating SimpleValue, then the serializer should print `{}`.
Here's a simple repo: https://github.com/GP4cK/built_value_absent

**Comment by davidmorgan** [comment]
I'm afraid that's not possible; the builder already uses `null` as the default, explicitly setting to `null` does not change the state so there is no way to make it serialize differently.

**Comment by GP4cK** [comment]
Thanks for your answer. Do you think that's a feature that could be developed? I think the package freezed managed to do it. although I haven't looked into the details yet. I'd be keen on helping.
Cheers.

**Comment by knaeckeKami** [comment]
I experimented somewhat successfully with this in ferry_generator, a code gen that generates graphql classes that use built_value.

At the moment I introduce a new `Value` type. This class just wraps any other value and is used to represent the following states:

- value is present and non-non null ( will serialize to  e.g. `"key" : "value"`  in json)
- value is present and null  ( e.g. `"key" : null` )
- value is absent (will not be serialized)

At the moment I generate a custom serializer for any type that has such a Value type, but I'd like to avoid that if possible since I feel this is tricky to get right in all corner cases and built_value has figured that out pretty well in the last 8 years ;).

Do you think it makes sense to add support for this directly in buit_value via the generated serializers?

Or maybe add a Plugin like StandardJsonPlugin that understand such a `Value` type and  wraps/unwraps these values?

**Comment by davidmorgan** [comment]
How does the `Value` clash distinguish the "present and null" case from the "absent" case, does it have an additional boolean field?

Possibly there could be support for such a boolean field indicating whether the null is present. I'd have to think about it though :)

**Comment by knaeckeKami** [comment]
At the moment I did it this way:

Value is just a wrapper around a nullable field:

```dart
class Value<T extends Object> {
  final T? _value;

  T? get value => _value;

  /// Create a (present) value by wrapping the [value] provided.
  const Value(T? value) : _value = value;

```
When there is an optional `String`field named `field`, it is wrapped in the following way:

```dart
Value<String>? get field
```

So now there are three possible states:

1. `field` itself is `null`, `field == null`
2. `field` is set to `Value(null)`. `field == Value(null)` 
3. `field` is set to some non-null String value

So I use the nullability of the wrapper to represent the absent state.
However, I just experimented with this yesterday and might come up with another implementation.

It works, but I'm not happy with it in its current form because it breaks the composability of nested builders and it requires custom serializers, so if I ship it, I'll probably come up with something else.

**Comment by knaeckeKami** [comment]
I have now released a first dev version of this feature. 
ferry_generator can now generate built_value classes which support differentiating between null and absent values.

This is done be wrapping each nullable field in a "Value" class

https://github.com/gql-dart/gql/blob/master/codegen/gql_tristate_value/lib/src/value.dart

This is a sealed class with two possible types, `PresentValue(value)` and `AbsentValue()`.

This allows us to represent three states:

- absent value which should not be serialized `const AbsentValue()`
- present value which is non-null `PresentValue("some value")`
- present value which is null `PresentValue(null)`

in order to make this work, each value -  field has to be initialized to `const AbsentValue()` in the `_initializeBuilder`.
Also, the class needs a custom Serializer which understands to Value type.

An example of Built-Class with value types and serializer can be found here:

https://github.com/gql-dart/gql/blob/master/codegen/end_to_end_test_tristate/lib/variables/__generated__/create_review.var.gql.dart

Is there any interest in adding a feature like this in built_value directly?

**Comment by davidmorgan** [comment]
Thanks Martin :)

Did I understand correctly that the Serializer of each class with a field of type `Value` needs modifying currently?

I wonder if you could implement this with a `SerializerPlugin`:

https://pub.dev/documentation/built_value/latest/serializer/SerializerPlugin-class.html

it gets called during serialization with information about the types, and can modify the data and response. So maybe it can make the changes needed for all types.

**Comment by knaeckeKami** [comment]
Yes, currently it generates a custom Serializer that handles wrapping/unwrapping  the `Value` and serializing only values that are wrapped in a PresentValue() type (if they are optional).

```dart
    if (_$episodevalue case _i1.PresentValue(value: final _$value)) {

```
